### PR TITLE
Make the index type of VectorContainer `itk::SizeValueType` by default

### PR DESCRIPTION
**Pull request by N-Dekker**
For most use cases, it appears that the index type of `itk::VectorContainer` should just be `itk::SizeValueType` (or `itk::IdentifierType`, but that's just an alias of `itk::SizeValueType`).

This pull request allows users to write `itk::VectorContainer<ElementType>` as a shorthand for `itk::VectorContainer<itk::SizeValueType, ElementType>`.

### CODE REVIEW

File: Modules/Core/SpatialObjects/include/itkEllipseSpatialObject.h [hunk 1]
```diff
@@ -100,13 +100,25 @@ class ITK_TEMPLATE_EXPORT EllipseSpatialObject : public SpatialObject<TDimension
...
 
 #if !defined(ITK_LEGACY_REMOVE)
-  itkLegacyMacro(void SetRadius(double radius)) { this->SetRadiusInObjectSpace(radius); }
-
-  itkLegacyMacro(void SetRadius(ArrayType radii)) { this->SetRadiusInObjectSpace(radii); }
-
-  itkLegacyMacro(ArrayType GetRadius() const) { return this->GetRadiusInObjectSpace(); }
-
-  itkLegacyMacro(void SetRadiiInObjectSpace(ArrayType radii)) { this->SetRadiusInObjectSpace(radii); }
+  itkLegacyMacro(void SetRadius(double radius))
```
Comment (dzenanz) [inline Modules/Core/SpatialObjects/include/itkEllipseSpatialObject.h:103]: This set of changes seems unrelated.

File: Modules/Core/SpatialObjects/include/itkMeshSpatialObject.h [hunk 2]
```diff
@@ -96,7 +96,10 @@ class ITK_TEMPLATE_EXPORT MeshSpatialObject : public SpatialObject<TMesh::PointD
 
 #if !defined(ITK_LEGACY_REMOVE)
   /** \deprecated Return the type of pixel used */
-  itkLegacyMacro(const char * GetPixelTypeName()) { return m_PixelType.c_str(); }
+  itkLegacyMacro(const char * GetPixelTypeName())
```
Comment (dzenanz) [inline Modules/Core/SpatialObjects/include/itkMeshSpatialObject.h:99]: Unrelated. Preferably group with the other into a separate commit.

File: Modules/Core/SpatialObjects/include/itkTubeSpatialObject.h [hunk 3]
```diff
@@ -85,7 +85,10 @@ class ITK_TEMPLATE_EXPORT TubeSpatialObject : public PointBasedSpatialObject<TDi
 
 #if !defined(ITK_LEGACY_REMOVE)
   /** Compute the tangents and normals of the centerline of the tube. */
-  itkLegacyMacro(bool ComputeTangentAndNormals()) { return ComputeTangentsAndNormals(); }
+  itkLegacyMacro(bool ComputeTangentAndNormals())
```
Comment (dzenanz) [inline Modules/Core/SpatialObjects/include/itkTubeSpatialObject.h:88]: Another unrelated change.

**Review by dzenanz — COMMENTED** [review]
Mostly looks good.

File: Modules/Core/SpatialObjects/include/itkEllipseSpatialObject.h (hunk 1 above)
Comment (N-Dekker) [reply Modules/Core/SpatialObjects/include/itkEllipseSpatialObject.h:103]: Thanks! Honestly I have no idea why clang-format decided to make those unrelated changes  🤷

File: Modules/Core/Common/include/itkVectorContainer.h [hunk 4]
```diff
@@ -44,10 +48,10 @@ namespace itk
  * \sphinxexample{Core/Common/IterateOnAVectorContainer,Iterate On A Vector Container}
  * \endsphinx
  */
-template <typename TElementIdentifier, typename TElement>
+template <typename T1, typename T2 = void>
 class ITK_TEMPLATE_EXPORT VectorContainer
   : public Object
-  , private std::vector<TElement>
+  , private std::vector<std::conditional_t<std::is_void_v<T2>, T1, T2>>
```
Comment (N-Dekker) [inline Modules/Core/Common/include/itkVectorContainer.h:54]: Unfortunately this "trick" does not work as smoothly as I was hoping for.... The compiler still considers `VectorContainer<ElementType>` and `VectorContainer<SizeValueType, ElementType>` to be different types. Which disallows interchanging one with the other. (For example, assigning a pointer to another.) Not really convenient 🤔

File: Modules/Core/SpatialObjects/include/itkEllipseSpatialObject.h (hunk 1 above)
Comment (N-Dekker) [reply Modules/Core/SpatialObjects/include/itkEllipseSpatialObject.h:103]: - @dzenanz Related pull request, doing those unrelated changes: #4857

Comment (N-Dekker) [reply Modules/Core/SpatialObjects/include/itkEllipseSpatialObject.h]: - Resolved by pull request #4857

File: Modules/Core/SpatialObjects/include/itkTubeSpatialObject.h (hunk 3 above)
Comment (N-Dekker) [reply Modules/Core/SpatialObjects/include/itkTubeSpatialObject.h]: - Resolved (again) by pull request #4857

File: Modules/Core/SpatialObjects/include/itkMeshSpatialObject.h (hunk 2 above)
Comment (N-Dekker) [reply Modules/Core/SpatialObjects/include/itkMeshSpatialObject.h]: - Resolved (again) by pull request #4857

**Review by dzenanz — APPROVED** [review]

File: Modules/Core/Common/include/itkVectorContainer.h [hunk 5]
```diff
@@ -21,11 +21,14 @@
...
 #include "itkObjectFactory.h"
 
+#include <type_traits> // For is_void_v.
 #include <utility>
 #include <vector>
 
 namespace itk
 {
+namespace ClassTemplate
+{
```
Comment (N-Dekker) [inline Modules/Core/Common/include/itkVectorContainer.h:31]: This pull request places the original class template `VectorContainer` in a nested namespace, `itk::TemplateClass`, so that it will still be available to users who _really need_ to access the original class. Usually, users may just use `itk::VectorContainer` (as before), which will then be an alias to `itk::TemplateClass::VectorContainer`.

This approach is a bit of a workaround. Maybe we can _eventually_ just define ``itk::VectorContainer<TElement>` as a template class, without the `TElement` parameter. But that would be a breaking change. 🤷

File: Modules/Core/Common/include/itkVectorContainer.h (hunk 4 above)
Comment (N-Dekker) [reply Modules/Core/Common/include/itkVectorContainer.h]: Resolved by making `itk::VectorContainer` an _alias_ _template_ of `itk::ClassTemplate::VectorContainer`!

File: Modules/Core/Common/include/itkVectorContainer.h [hunk 6]
```diff
@@ -561,14 +564,30 @@ class ITK_TEMPLATE_EXPORT VectorContainer
...
+ * The template parameters `T1` and `T2` allow specifying the index type and the element type, as follows:
+ *
+ * \tparam T1 The index type OR (when `T2` is `void`) the element type.
+ *
+ * \tparam T2 The element type OR `void`. When `T2` is `void`, the element type is specified by the first template
+ * argument (T1), and the index type will be `SizeValueType`.
+ */
+template <typename T1, typename T2 = void>
+using VectorContainer = ClassTemplate::VectorContainer<std::conditional_t<std::is_void_v<T2>, SizeValueType, T1>,
+                                                       std::conditional_t<std::is_void_v<T2>, T1, T2>>;
```
Comment (thewtex) [inline Modules/Core/Common/include/itkVectorContainer.h:582]: Cool!

File: Modules/Core/Common/include/itkVectorContainer.h (hunk 5 above)
Comment (thewtex) [reply Modules/Core/Common/include/itkVectorContainer.h:31]: I think we should use `itk::detail::` here instead of creating a new namespace.

**Review by thewtex — CHANGES_REQUESTED** [review]
@N-Dekker awesome!!

Naming comment inline.

File: Modules/Core/Common/include/itkVectorContainer.h (hunk 5 above)
Comment (N-Dekker) [reply Modules/Core/Common/include/itkVectorContainer.h:31]: Thanks for the suggestion @thewtex. Can you please explain a little bit? What is the general rule for _when_ to place something in `itk::detail::`?

**Review by dzenanz — APPROVED** [review]

**Review by thewtex — APPROVED** [review]
Excellent!! :sparkles:
